pkg/util/constraint/constraint_ha_status.go

```diff
@@ -0,0 +1,57 @@
+package constraint
+
+import (
+	"encoding/json"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// Error represents a single error caught while adding a constraint to OPA
+type Error struct {
+	Code     string `json:"code"`
+	Message  string `json:"message"`
+	Location string `json:"location,omitempty"`
+}
+
+// ByPodStatus defines the observed state of a constraint as seen by
+// an individual controller
+type ByPodStatus struct {
+	// a unique identifier for the pod that wrote the status
+	ID                 string  `json:"id,omitempty"`
+	ObservedGeneration int64   `json:"observedGeneration,omitempty"`
+	Errors             []Error `json:"errors,omitempty"`
+	Enforced           bool    `json:"enforced,omitempty"`
+}
+
+func GetHAStatus(obj *unstructured.Unstructured) (*ByPodStatus, error) {
+	s, err := util.GetHAStatus(obj)
```

Comment:
I believe you can now move the (Get/Set/Delete)HAStatus functions out of util/ha_status.go since they are only really called from here?

Reply:
It's technically possible, but they still have value as unstructured-type functions. Also, if there are other resources in the future that require by-pod status, they will (like the constraint util set) rely on this common logic for extracting the status.

Reply:
In general, I would favor YAGNI principle and make these common functions when you have concrete need for them. Just a suggestion though.

Reply:
I can move them here and hide them for now, but keep the separation of concerns in case it's needed later.